ethereum/db/overlay_db.py

```diff
@@ -0,0 +1,61 @@
+from ethereum import utils
+from rlp.utils import str_to_bytes
+
+from .base_db import BaseDB
+
+# Used for making temporary objects
+class OverlayDB(BaseDB):
+
+    def __init__(self, db):
+        self.db = db
+        self.kv = None
+        self.overlay = {}
+
+    def get(self, key):
+        if key in self.overlay:
+            if self.overlay[key] is None:
+                raise KeyError()
+            return self.overlay[key]
+        return self.db.get(key)
+
+    def put(self, key, value):
+        self.overlay[key] = value
+
+    def delete(self, key):
+        self.overlay[key] = None
```

Comment:
Ah.  I had missed that.

On Thu, Apr 7, 2016, 10:12 PM Jan Xie <notifications@github.com> wrote:

> In ethereum/db/overlay_db.py
> <https://github.com/ethereum/pyethereum/pull/332#discussion_r58978299>:
>
> > +        self.db = db
> > +        self.kv = None
> > +        self.overlay = {}
> > +
> > +    def get(self, key):
> > +        if key in self.overlay:
> > +            if self.overlay[key] is None:
> > +                raise KeyError()
> > +            return self.overlay[key]
> > +        return self.db.get(key)
> > +
> > +    def put(self, key, value):
> > +        self.overlay[key] = value
> > +
> > +    def delete(self, key):
> > +        self.overlay[key] = None
>
> If deleted, get will read through to underlying db, see line#16.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/ethereum/pyethereum/pull/332/files/e59db9cee6abcf4bfd93a3a941fcf88355979e7a#r58978299>
>